ocaml/xapi-storage-script/main.ml

```diff
@@ -1739,46 +1743,93 @@ let bind ~volume_script_dir =
   S.DATA.MIRROR.receive_cancel (u "DATA.MIRROR.receive_cancel") ;
   S.SR.update_snapshot_info_src (u "SR.update_snapshot_info_src") ;
   S.DATA.MIRROR.stop (u "DATA.MIRROR.stop") ;
-  Rpc_async.server S.implementation
+  Rpc_lwt.server S.implementation
 
 let process_smapiv2_requests server txt =
   let request = Jsonrpc.call_of_string txt in
-  server request >>= fun response ->
-  Deferred.return (Jsonrpc.string_of_response response)
+  let to_err e =
+    Storage_interface.(rpc_of Errors.error Errors.(Internal_error e))
+  in
+  Lwt.try_bind
+    (fun () -> server request)
+    (fun response -> Lwt.return (Jsonrpc.string_of_response response))
+    (fun exn ->
+      Printexc.to_string exn |> to_err |> Jsonrpc.to_string |> Lwt.return
+    )
 
 (** Active servers, one per sub-directory of the volume_root_dir *)
-let servers = String.Table.create () ~size:4
+let servers = Base.Hashtbl.create ~size:4 (module Base.String)
 
 (* XXX: need a better error-handling strategy *)
-let get_ok = function
-  | Ok x ->
-      x
-  | Error e ->
-      let b = Buffer.create 16 in
-      let fmt = Format.formatter_of_buffer b in
-      Message_switch_unix.Protocol_unix.Server.pp_error fmt e ;
-      Format.pp_print_flush fmt () ;
-      failwith (Buffer.contents b)
+let get_ok x =
+  Message_switch_unix.Protocol_unix.Server.error_to_msg x
+  |> Result.fold ~ok:Fun.id ~error:(function `Msg err -> failwith err)
 
 let rec diff a b =
```

Comment:
Could this just be to use `List.filter` like: `List.filter Fun.(negate ((flip List.mem) b)) a`?

Reply:
That would be one way to implement it, yes. A Set would be more efficient, though